velox/exec/tests/HashJoinTest.cpp

```diff
@@ -2876,6 +2876,42 @@ TEST_P(MultiThreadedHashJoinTest, noSpillLevelLimit) {
       .run();
 }
 
+TEST_F(HashJoinTest, duplicatedJoinkeys) {
+  auto probeVectors = makeBatches(3, [&](int32_t /*unused*/) {
+    return makeRowVector({
+        makeFlatVector<int64_t>({1, 2, 2, 3, 3, 3, 4, 5, 5, 6, 7}),
+        makeFlatVector<int64_t>({1, 2, 2, 3, 3, 3, 4, 5, 5, 6, 8}),
+    });
+  });
+
+  auto buildVectors = makeBatches(3, [&](int32_t /*unused*/) {
+    return makeRowVector({
+        makeFlatVector<int64_t>({1, 1, 3, 4, 5, 5, 7, 8}),
+    });
+  });
+
+  createDuckDbTable("t", probeVectors);
+  createDuckDbTable("u", buildVectors);
+
+  auto planNodeIdGenerator = std::make_shared<core::PlanNodeIdGenerator>();
+  VELOX_ASSERT_THROW(
+      PlanBuilder(planNodeIdGenerator)
+          .values(probeVectors)
+          .project({"c0 AS t0", "c1 as t1"})
+          .hashJoin(
+              {"t0", "t1"},
+              {"u0", "u0"},
```

Comment:
@rui-mo Rui, I may have mis-understood the use case. I was thinking it is SELECT * FROM t, u WHERE t.key = u.key AND t.key = u.key (a pair of duplicate join keys), but looks like you have a difference case: SELECT * FROM t, u WHERE t.key1 = u.key AND t.key2 = u.key. Is this so?

Are you changing the plan to add a filter for 't' before the join?

- Join t.key1 = u.key
    - Filter t.key1 = t.key2
        - TableScan t
     - TableScan u

Reply:
Yes, that's the case. This is because Spark allows that behavior and generates such join, but indeed, we need to change the plan if Velox rejects it.

Reply:
It is obscure to me which is better on performance. By adding a filter, join can be simplified and improved, but the filter requires some time for execution. If this kind of plan is functionally needed in Velox in some cases, I can continue to work on it in this PR.

Reply:
For non-inner joins, this would the question of correctness, not performance. It may not be possible to rewrite these joins using a filter. Any chance you could create a separate PR with your original change?

Reply:
Sure. I'll add more tests to verify left/right/outer join.

Reply:
@mbasmanova Opened https://github.com/facebookincubator/velox/pull/6084. Could you take a look? I find in Velox, full outer join is supported with NestedLoopJoin so its test is not added.

Reply:
@rui-mo I'm out next week. Can you ping me on the following Monday?

BTW, NestedLoopJoin should only be used when there is no equi clause for the join.

Reply:
@mbasmanova Do you have time to review https://github.com/facebookincubator/velox/pull/6084 again? Tests for full outer is also added.